File: js-template/leetcode/editor/cn/find-all-anagrams-in-a-string.js
/*
 * @lc app=leetcode.cn id=438 lang=javascript
 * @lcpr version=30404
 *
 * [438] 找到字符串中所有字母异位词
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
    var result = [];
    
    const need = new Map();
    for (let i = 0; i < p.length; i++) {
        const char = p[i];
        need.set(char, (need.get(char) || 0) + 1);
    }
    const window = new Map();

    let left = 0, right = 0;
    let valid = 0;
    
    while (right < s.length) {
        const c = s[right];
        right++;

        if (need.has(c)) {
            window.set(c, (window.get(c) || 0) + 1);
            if (window.get(c) === need.get(c))
                valid++;
        }

        while(right - left >= p.length) {
            if (valid === need.size) {
                result.push(left);
            }
            let d = s[left];
            left++;
            if (need.has(d)) {
                if (window.get(d) === need.get(d))
                    valid--;
            }
            window.set(d, window.get(d) - 1);
        }
    }

    return result;
};
// @lc code=end

// your test code here

console.log(findAnagrams("cbaebabacd", "abc")) // [0, 7];



/*
// @lcpr case=start
// "cbaebabacd"\n"abc"\n
// @lcpr case=end

// @lcpr case=start
// "abab"\n"ab"\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/find-all-anagrams-in-a-string.cpp
/*
 * @lc app=leetcode.cn id=438 lang=cpp
 * @lcpr version=30404
 *
 * [438] 找到字符串中所有字母异位词
 */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> result;

        unordered_map<char, int> need;
        unordered_map<char, int> window;
        for (auto &&c : p)
            need[c]++;

        int left = 0, right = 0;
        int valid = 0;

        while (right < s.size())
        {
            char c = s[right];
            right++;

            if (need.contains(c)) {
                window[c]++;
                if (window[c] == need[c])
                    valid++;
            }

            while (right - left >= p.size())
            {
                if (valid == need.size())
                    result.push_back(left);
                
                char d = s[left];
                ++left;
                
                if (need.contains(d)) {
                    if (window[d] == need[d])
                        valid--;
                    window[d]--;
                }
            }
        }
        return result;
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    solution.findAnagrams("cbaebabacd", "abc");
}



/*
// @lcpr case=start
// "cbaebabacd"\n"abc"\n
// @lcpr case=end

// @lcpr case=start
// "abab"\n"ab"\n
// @lcpr case=end

 */

